### ISSUE TITLE
PeakFileOptions should not be set as member of FileHandler

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#7168]
Some design discussion for the next refactoring of the FileIO:

Having the PeakFileOption as part of FileHandler looks like it is not the correct/final design.
```
    FileHandler f;
    PeakFileOptions options;
    options.addMSLevel(1);
    f.getOptions() = options;
    f.loadExperiment(in_mzml, unprocessed_spectra, {FileTypes::MZML}, log_type_);
```

A standard OOP way would be to use a factory pattern:
```
    PeakFileLoaderInterface* p = FileHandler::getPeakFileLoader({FileTypes::MZML});
    PeakFileOptions options;
    options.addMSLevel(1);
    p->setOptions(options);
    p->load(in_mzml, unprocessed_spectra, log_type_);
```

@cbielow what is your take on this?

**Comment by poshul** [comment]
The  factory option certainly seems neater. I am curious if there would be
a downside to just making the fileHandler static, and passing options as
part of the load or store calls. The majority of calls to fileHandler now
just create the object, use it once and then discard it.

On Tuesday, October 31, 2023, Timo Sachsenberg ***@***.***>
wrote:
> Assigned #7168 to @poshul.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.<
https://ci4.googleusercontent.com/proxy/elt_flPW3JKd2lCl2gUV9gCPQ1akVeyfj3V5a2_DgCniZHARzH4nR7NxNkH7kj7db5AQ4q1w2yqwJmcPBSh7d8JeE-JUwj0Ie4XegLPlP5CaDbOCEaMhmls4XjFMKqsm4FVJuPZiTJNzul61oaKIWiCeXgwjVlh2LJrZl0ZvhHuuIbo4FhcGDVMOvaB4OEpj90md6WvU5syXHTbVKdzwWTYn12UVsixrrXX2b5c_FUVEH5imRRkhe67js1wDS8H1WB1yWS6N2FPOMirO=s0-d-e1-ft#https://github.com/notifications/beacon/AABD3C56JWUJQVEDZZZKHQDYCDBYDA5CNFSM6AAAAAA6XMESD6WGG33NNVSW45C7OR4XAZNWJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XKUY3PNVWWK3TUL5UWJTYAAAAAFBGQGEZQ.gif>Message
ID: ***@***.***>

**Comment by timosachsenberg** [comment]
Yeah I think this would be already an improvement that could make later refactoring easier. 
One thing we need to think through is how to handle a call the generic calls you are working towards (e.g., if you say you want to load a given filename, and it must give you an MSExperiment).
E.g., how to handle if some potential file formats (raw, mzML, DTA) don't support the peak options or not all of them.
```
 f.loadExperiment(in_mzml, unprocessed_spectra, {}, log_type_); // not sure if this is the correct syntax but basically allow everything that gives you an MSExperiment
```

I am thinking about how to make sure that e.g. a TOPP tools knows about which formats it can load (as it needs to set its valid file formats).

Something like:
```
PeakFileLoaderInterface* p = FileHandler::getPeakFileLoader({}, options);
auto supported_file_extensions = p->getSupportedExtensions();
registerInputFile("in", ...) 
setValidFormats("in", supported_file_extensions)
```
could in principle work.
